### PR TITLE
[SPARK-43031] [SS] [Connect] Enable unit test and doctest for streaming

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Enable unit tests and doctests for streaming queries. A lot are skipped and needs to be un-skipped as the development goes on. 

Note that I also separated the `{foreach, foreachBatch}` tests from the original test suite. Because currently they are not implemented in connect and it seems unnecessary to manually add a skip for all of them in `StreamingParityTests`. Also it doesn't hurt to separate them I think as these tests are large enough already.

### Why are the changes needed?

More tests is always better than less.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

It's test itself.

### CODE REVIEW

File: python/pyspark/sql/connect/streaming/readwriter.py [hunk 1]
```diff
@@ -168,9 +168,75 @@ def json(
```
Comment (WweiL) [inline python/pyspark/sql/connect/streaming/readwriter.py:168]: Please ignore this change. They are added in https://github.com/apache/spark/pull/40689.

File: python/pyspark/sql/streaming/readwriter.py [hunk 2]
```diff
@@ -1150,6 +1150,7 @@ def foreach(self, f: Union[Callable[[Row], None], "SupportsProcess"]) -> "DataSt
         This API is evolving.
 
         Examples
+        TODO(SPARK-43054): remove the SKIP flags below
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/readwriter.py:1153]: no biggie but should probably put this comment on the top of this method .. otherwise it'll be shown in API reference documentation ..

**Comment by WweiL** [comment]
CC @rangadi @pengzhon-db

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+#
+
+import os
+import tempfile
+import time
+
+from pyspark.testing.sqlutils import ReusedSQLTestCase
+
+
+class StreamingTestsForeachFamilyMixin:
```
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:25]: Wondering why 'foreach family'. What are the other foreach() methods tested here?

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    class ForeachWriterTester:
```
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:26]: When I tried these tests, I got a serialization error after moving this to 'Mixin'. We could handle it when run into it with connect tests.

File: python/pyspark/sql/tests/streaming/test_streaming.py [hunk 5]
```diff
@@ -77,8 +109,14 @@ def test_stream_read_options_overwrite(self):
                 .schema(bad_schema)
                 .load(path="python/test_support/sql/streaming", schema=schema, format="text")
             )
-        self.assertTrue(df.isStreaming)
-        self.assertEqual(df.schema.simpleString(), "struct<data:string>")
+            # TODO: Moving this outside of with block will trigger the following error,
```
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming.py:112]: It does not need to be a TODO. Testing it inside `with` is the right thing.
Calls like `schema` are executed on server with the updated state of the conf.

File: python/pyspark/sql/tests/streaming/test_streaming.py [hunk 6]
```diff
@@ -26,7 +26,39 @@
 from pyspark.testing.sqlutils import ReusedSQLTestCase
 
 
-class StreamingTests(ReusedSQLTestCase):
+class StreamingTestsMixin:
+    def test_streaming_query_functions_sanity(self):
```
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming.py:30]: Is this a new test?
Could you add brief description? `test... sanity` does not sound good. May be `test_streaming_query_functions_basic()` ?

File: python/pyspark/sql/tests/streaming/test_streaming.py [hunk 7]
```diff
@@ -295,334 +333,6 @@ def test_query_manager_await_termination(self):
             q.stop()
             shutil.rmtree(tmpPath)
 
-    class ForeachWriterTester:
```
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming.py:298]: Moving foreach tests to a different files SGTM.

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py (hunk 3 above)
Comment (rangadi) [inline python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:25]: Better not to have Mixin class yet. We don't have corresponding '_partity_' tests in connect for this.

File: python/pyspark/sql/streaming/query.py [hunk 8]
```diff
@@ -188,7 +192,7 @@ def awaitTermination(self, timeout: Optional[int] = None) -> Optional[bool]:
 
         Return whether the query has terminated or not within 5 seconds
 
-        >>> sq.awaitTermination(5)
+        >>> sq.awaitTermination(5) # doctest: +SKIP
```
Comment (rangadi) [inline python/pyspark/sql/streaming/query.py:195]: Why are these needed? This won't be tested in connect yet, right?

File: python/pyspark/sql/streaming/readwriter.py [hunk 9]
```diff
@@ -89,8 +92,8 @@ def format(self, source: str) -> "DataStreamReader":
 
         Examples
         --------
-        >>> spark.readStream.format("text")
-        <pyspark.sql.streaming.readwriter.DataStreamReader object ...>
+        >>> spark.readStream.format("text") # doctest: +ELLIPSIS
```
Comment (rangadi) [inline python/pyspark/sql/streaming/readwriter.py:95]: Is `# doctest: +ELLIPSIS` needed? I didn't add in my PR (#40586)

**Review by rangadi — COMMENTED** [review]
Overall LGTM. Made a few comments.

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py (hunk 3 above)
Comment (WweiL) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:25]: Right I put both `foreach` and `foreachBatch` to this class. My naming sense isn't that great so I'm very welcome to any suggestion on it's name here...

Comment (WweiL) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:25]: Sure I'll remove the mixin. Still better to create this class though, to avoid tons of `@unittest.skip` in the streaming parity test class.

File: python/pyspark/sql/tests/streaming/test_streaming.py (hunk 5 above)
Comment (WweiL) [reply python/pyspark/sql/tests/streaming/test_streaming.py:112]: I see. So this is a behavior change for migrating to connect?

File: python/pyspark/sql/streaming/readwriter.py (hunk 9 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/readwriter.py:95]: Hmm I believe you added these? https://github.com/apache/spark/pull/40586/files#r1157045341

File: python/pyspark/sql/streaming/query.py (hunk 8 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/query.py:195]: Right.. This actually won't be tested in both connect and non-connect I believe (@HyukjinKwon please correct me if I'm wrong) , just some temporary pain we have to bear with

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py (hunk 4 above)
Comment (WweiL) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:26]: I get all tests passed by running ` python/run-tests --testnames pyspark.sql.tests.streaming.test_streaming_foreach_family`. I renamed the mixin part though in the new commit

File: python/pyspark/sql/streaming/readwriter.py (hunk 9 above)
Comment (rangadi) [reply python/pyspark/sql/streaming/readwriter.py:95]: I didn't add. E.g.:
image

File: python/pyspark/sql/streaming/query.py (hunk 8 above)
Comment (rangadi) [reply python/pyspark/sql/streaming/query.py:195]: What is the pain? What fails if you remove this diff?

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py (hunk 4 above)
Comment (rangadi) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:26]: Can we move `foreach` back to its original place. Foreach() and foreachbath() are very different. The later takes a dataframe, the former is more like a UDF.

Comment (rangadi) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py:26]: Please add foreach() api to one of the jiras or a new jira.

File: python/pyspark/sql/streaming/readwriter.py (hunk 9 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/readwriter.py:95]: Ah I see, in line 538 there is one. I could remove all of the `# doctest: +ELLIPSIS` as we already defined them in the option: https://github.com/apache/spark/blob/17720b7121eee050a546126fad7ce8229fe6bda2/python/pyspark/sql/streaming/readwriter.py#L1504

https://stackoverflow.com/questions/17092215/how-enable-ellipsis-when-calling-python-doctest

File: python/pyspark/sql/streaming/query.py [hunk 10]
```diff
@@ -330,7 +335,8 @@ def stop(self) -> None:
         Stop streaming query
 
         >>> sq.stop()
-        >>> sq.isActive
+
+        >>> sq.isActive # doctest: +SKIP
```
Comment (rangadi) [inline python/pyspark/sql/streaming/query.py:339]: is this required? Why is it skipped?

File: python/pyspark/sql/streaming/readwriter.py (hunk 9 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/readwriter.py:95]: Actually maybe it doesn't hurt to put them there..? We already have this flag set, e.g.:  https://github.com/apache/spark/blob/master/python/pyspark/sql/streaming/query.py#L222

File: python/pyspark/sql/streaming/query.py (hunk 10 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/query.py:339]: Calling isActive after the query is stopped will throw error right now before the better session management is implemented

Comment (rangadi) [reply python/pyspark/sql/streaming/query.py:339]: Is see, please add a TODO here with a brief comment.

File: python/pyspark/sql/streaming/readwriter.py (hunk 9 above)
Comment (rangadi) [reply python/pyspark/sql/streaming/readwriter.py:95]: No need to add them. It makes it the test less readable.

File: python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py (hunk 3 above)
Comment (rangadi) [reply python/pyspark/sql/tests/streaming/test_streaming_foreach_family.py]: Class is fine.

File: python/pyspark/sql/streaming/query.py (hunk 8 above)
Comment (rangadi) [reply python/pyspark/sql/streaming/query.py:194]: Lets not skip any tests. Any skip should have a TODO comment.

Comment (WweiL) [reply python/pyspark/sql/streaming/query.py:194]: Sure! I've moved the TODO to the top of these methods to avoid them showing up in the PySpark doc

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming/query.py:194]: If we're absolutely sure that we will enable this back very soon, I am fine. As @rangadi said, I tried my best to avoid skipping the tests, but I did few times when I am about to enable it back very soon.

File: python/pyspark/sql/streaming/readwriter.py [hunk 11]
```diff
@@ -1359,7 +1364,8 @@ def start(
         >>> q.name
         'this_query'
         >>> q.stop()
-        >>> q.isActive
+
```
Comment (pengzhon-db) [inline python/pyspark/sql/streaming/readwriter.py:1367]: nit: remove the space?

**Review by HyukjinKwon — APPROVED** [review]
looks fine from a cursory look

File: python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+# limitations under the License.
+#
+
+import unittest
+
+from pyspark.sql.tests.streaming.test_streaming import StreamingTestsMixin
+from pyspark.testing.connectutils import ReusedConnectTestCase
+
+
+class StreamingParityTests(StreamingTestsMixin, ReusedConnectTestCase):
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py:24]: uhoh, the tests here shouldn't be skipped.

[the remark this answers is not in the document]
Comment (WweiL) [reply python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py:24]: This file marks all tests that should be skipped currently because these functionality is not supported yet in connect.

It executes all tests from `StreamingTestsMixin`, except these below marked skipped.

When anyone adds support to corresponding functions, they should delete the @unittest.skip and function, unless they want to do some change specific for connect.

So ideally in the end, this class should only be like 
```
class StreamingParityTests(StreamingTestsMixin, ReusedConnectTestCase):
    pass
```
And then it runs all tests from StreamingTestsMixin

File: python/pyspark/sql/streaming/query.py (hunk 8 above)
Comment (WweiL) [reply python/pyspark/sql/streaming/query.py]: Thanks! Done

File: python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rangadi) [reply python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py:24]: Each of the skipped tests here has a SPARK ticket associated with it. We will enable these soon. LGTM.

**Review by rangadi — APPROVED** [review]
LGTM. @WweiL ping Hyukjin when this is ready to be merged.

**Review by pengzhon-db — APPROVED** [review]

**Comment by WweiL** [comment]
@HyukjinKwon can you merge this? Thank you!

**Comment by HyukjinKwon** [comment]
Merged to master.
